### PR TITLE
[K9VULN-2298] feat: detect direct and indirect dependencies in autogenerated requirements.txt files

### PR DESCRIPTION
**Pull request by amaanq**
### Problem

Currently, we do not bother to determine if a dependency in a `requirements.txt` file is direct or indirect. This is a concern because users see their dependencies detected from these files showing up as transitive, even though they're direct, which is not great. As such, the proposed change here follows a proposal where we attempt to parse the comments in a given `requirements.txt` file. When using `pip-compile`, comments are added at the beginning of the file stating that this is an autogenerated file, and following the dependencies that indicate why this dependency was listed. It can be because it was specified in a `requirements.in` file, which is manually edited, or because it is a dependency of another package.

### Solution

First, we mark all dependencies as direct for requirements.txt files. The reasoning behind this is twofold - one, it's better to assume the user manually wanted to add it, rather than assume they're all transitive, which is oftentimes not the case, and two, we are attempting to figure out indirect ones *if* there is such a comment that lets us know this. If there isn't, we're assuming this was manually created then.

Second, we attempt to parse the comments in a given `requirements.txt` file *iff* the file was detected to be machine generated. The heuristic for determining if a given file is machine generated is by peeking at the first group of comments. If, inside this group of comments the word `autogenerated` is present, we assume it's machine-generated by `pip-compile` and parse comments as they appear. Otherwise, we would assume it's manually created, and every dependency is a direct dependency. When we parse the comments in the autogenerated file, we detect if the comment starts with via. If it does, we look for either a single package name (in the format of 1 or more alphanumeric or `-` or `_` characters) followed by the end of the comment, or `-r`. In the former case, this implies that the current dependency is an indirect dependency of another one in the project. In the latter case, this implies that the dependency is a direct dependency

### Testing

There are unit tests and integration tests that test parsing behavior for a variety of cases, including the comment being split up on multiple lines, as well as checking for user-comments that happen to start with `via`, and testing comments at the end of the file.

### CODE REVIEW

File: pkg/lockfile/parse-requirements-txt_test.go [hunk 1]
```diff
@@ -235,6 +236,7 @@ func TestParseRequirementsTxt_MultipleRequirementsConstrained(t *testing.T) {
 				Filename: path,
 			},
 			DepGroups: []string{"multiple-packages-constrained"},
+			IsDirect:  true,
```
Comment (juli1) [inline pkg/lockfile/parse-requirements-txt_test.go:239]: For example, here, this should not be true.

**Review by juli1 — CHANGES_REQUESTED** [review]
Please check the test.

Let's take for example, with `multiple-packages-constrained.txt`:
 - `beautifulsoup` should be transitive
 - `astroid` should be transitive


Please have better test cases and ensure the logic works and is solid.

**Comment by amaanq** [comment]
Tests have been updated to include the generated tidbit at the top that `pip-compile` outputs

File: cmd/osv-scanner/fixtures/encoding-integration-test-locks/UTF-8/requirements.txt [hunk 2]
```diff
@@ -1,7 +1,29 @@
+#
```
Comment (marcwieserdev) [inline cmd/osv-scanner/fixtures/encoding-integration-test-locks/UTF-8/requirements.txt:1]: Can you have 2 requirements.txt files in that test, one manually generated with multiple file links (usage of the -r flag) and one automatically generated from a requirements.in ?

**Review by marcwieserdev — COMMENTED** [review]
We're missing some tests from manual files (as the existing ones has been transformed to auto-generated).

We should have them as it is one of the usage of requirement.txt file, especially some with comments, with usage of -r flag to link to other files.

File: cmd/osv-scanner/fixtures/encoding-integration-test-locks/UTF-8/requirements.txt (hunk 2 above)
Comment (amaanq) [reply cmd/osv-scanner/fixtures/encoding-integration-test-locks/UTF-8/requirements.txt:1]: good call, added another file (requirements-test.txt) that is not autogenerated with a link to another requirements file

**Review by juli1 — APPROVED** [review]
the test and logic look good to me, please address @marcwieserdev feedback before merging.

**Review by marcwieserdev — COMMENTED** [review]
Seems good, thanks 🙏

**Review by marcwieserdev — APPROVED** [review]

**Pull request merged**